lib/sycamore/sycamore/tests/integration/connectors/opensearch/test_opensearch_read.py

```diff
@@ -110,3 +115,89 @@ def test_ingest_and_read(self, setup_index, exec_mode):
+            import pickle
+
+            data = pickle.load(open(f"{cache_dir}/{found}", "rb"))
+            return Document(**data)
+
+        path = str(TEST_DIR / "resources/data/pdfs/Ray.pdf")
+        context = sycamore.init(exec_mode=exec_mode)
+        original_docs = (
+            context.read.binary(path, binary_format="pdf")
+            .partition(partitioner=UnstructuredPdfPartitioner())
+            .materialize(cache_dir)
+            .explode()
+            .write.opensearch(
+                os_client_args=TestOpenSearchRead.OS_CLIENT_ARGS,
+                index_name=TestOpenSearchRead.INDEX,
+                index_settings=TestOpenSearchRead.INDEX_SETTINGS,
+                execute=False,
+            )
+            .take_all()
+        )
+
+        with OpenSearch(**TestOpenSearchRead.OS_CLIENT_ARGS) as os_client:
+            os_client.indices.refresh(TestOpenSearchRead.INDEX)
+
+        retrieved_docs = context.read.opensearch(
+            os_client_args=TestOpenSearchRead.OS_CLIENT_ARGS, index_name=TestOpenSearchRead.INDEX
+        )
+        target_doc_id = original_docs[-1].doc_id if original_docs[-1].doc_id else ""
+        query = {"query": {"term": {"_id": target_doc_id}}}
+        query_docs = context.read.opensearch(
+            os_client_args=TestOpenSearchRead.OS_CLIENT_ARGS, index_name=TestOpenSearchRead.INDEX, query=query
+        )
+
+        retrieved_docs_reconstructed = context.read.opensearch(
+            os_client_args=TestOpenSearchRead.OS_CLIENT_ARGS,
+            index_name=TestOpenSearchRead.INDEX,
+            reconstruct_document=True,
+            doc_reconstructor=doc_reconstructor,
+        )
+        original_materialized = sorted(original_docs, key=lambda d: d.doc_id)
+
+        # hack to allow opensearch time to index the data. Without this it's possible we try to query the index before
+        # all the records are available
+        time.sleep(1)
+        retrieved_materialized = sorted(retrieved_docs.take_all(), key=lambda d: d.doc_id)
+        query_materialized = query_docs.take_all()
+        retrieved_materialized_reconstructed = sorted(retrieved_docs_reconstructed.take_all(), key=lambda d: d.doc_id)
+
+        # with OpenSearch(**TestOpenSearchRead.OS_CLIENT_ARGS) as os_client:
+        #    os_client.indices.delete(TestOpenSearchRead.INDEX)
+        assert len(query_materialized) == 1  # exactly one doc should be returned
+        compare_connector_docs(original_materialized, retrieved_materialized)
+
+        assert len(retrieved_materialized_reconstructed) == 1
+        doc = retrieved_materialized_reconstructed[0]
+        assert len(doc.elements) == len(retrieved_materialized) - 1  # drop the document parent record
+
+        for i in range(len(doc.elements) - 1):
+            assert doc.elements[i].element_index < doc.elements[i + 1].element_index
+
+        # Clean slate between Execution Modes
+        with OpenSearch(**TestOpenSearchRead.OS_CLIENT_ARGS) as os_client:
```

Comment:
Make the os_client once rather than multiple times?

Reply:
Ok